fairseq2n/src/fairseq2n/data/image/image_decoder.cc

```diff
@@ -0,0 +1,204 @@
+        throw_<std::invalid_argument>(
+            "The input data must be of type `memory_block`, but is of type `{}` instead.", d.type());
+    
+    const memory_block &block = d.as_memory_block();
+    if (block.empty())
+        throw_<std::invalid_argument>(
+            "The input memory block has zero length and cannot be decoded.");
+
+    auto data_ptr = block.data();
+    data output;
+ 
+    const std::array<uint8_t, 3> jpeg_signature = {255, 216, 255};
+    const std::array<uint8_t, 4> png_signature = {137, 80, 78, 71};
+
+    if(memcmp(jpeg_signature.data(), data_ptr, 3) == 0) {
+        output = decode_jpeg(block);
+    } else if(memcmp(png_signature.data(), data_ptr, 4) == 0) {
+        output = decode_png(block);
+    } else {
+        throw_<std::invalid_argument>(
+            "Unsupported image file. Only jpeg and png are currently supported.");
+    }
+    return output;
+}
+
+data
+image_decoder::decode_png(const memory_block &block) const 
+{
+    png_structp png_ptr = png_create_read_struct(PNG_LIBPNG_VER_STRING, nullptr, nullptr, nullptr);
+    if (png_ptr == nullptr) {
+        throw_<internal_error>("Failed to create PNG read struct.");
+    }
+    png_infop info_ptr = png_create_info_struct(png_ptr);
+    if (info_ptr == nullptr) {
+        png_destroy_read_struct(&png_ptr, nullptr, nullptr);
+        throw_<internal_error>("Failed to create PNG info struct.");
+    }
+
+    auto data_ptr = png_const_bytep(block.data());
+    auto data_len = block.size();
+
+    struct Reader {
+    png_const_bytep ptr;
+    png_size_t count;
+    Reader(png_const_bytep p, png_size_t c) : ptr(p), count(c) {}
+    };
+    Reader reader(data_ptr + 8, data_len - 8);
+    
+    auto read_callback = [](png_structp png_ptr2,
+                          png_bytep output,
+                          png_size_t bytes) {
+    auto reader = static_cast<Reader*>(png_get_io_ptr(png_ptr2));
+    std::copy(reader->ptr, reader->ptr + bytes, output);
+    reader->ptr += bytes;
+    reader->count -= bytes;
+    };
+
+    png_set_sig_bytes(png_ptr, 8);
+    png_set_read_fn(png_ptr, &reader, read_callback);
+    png_read_info(png_ptr, info_ptr);
+
+    png_uint_32 width=0, height=0;
+    int bit_depth=0, color_type=0;
+    int interlace_type=0;
+    auto retval = png_get_IHDR(
+        png_ptr,
+        info_ptr,
+        &width,
+        &height,
+        &bit_depth,
+        &color_type,
+        &interlace_type,
+        nullptr,
+        nullptr);
+
+    if (retval != 1) {
+        png_destroy_read_struct(&png_ptr, &info_ptr, nullptr);
+        throw_<std::invalid_argument>("Could not read image metadata from content.");
+    }
+
+    if (is_little_endian()) {
+      png_set_swap(png_ptr);
+    }
+    int channels = png_get_channels(png_ptr, info_ptr);
+
+    at::ScalarType dtype = bit_depth <= 8 ? at::kByte : at::kShort;
+    at::Tensor image = at::empty({height, width, channels}, at::dtype(dtype).device(at::kCPU).pinned_memory(opts_.pin_memory()));
```

Comment:
For instance, if this statement throws an error, you will leak `png_ptr` since `png_destroy_read_struct` won't be called. The technique used in C++ to handle this sort of cases is called RAII (Resource Acquisition is Initialization) (see [here](https://en.cppreference.com/w/cpp/language/raii) for a formal description). You can find a simple RAII class for "file descriptors" [here](https://github.com/facebookresearch/fairseq2/blob/main/fairseq2n/src/fairseq2n/data/detail/file.h#L20) in our code base. In fact both `fairseq2n::detail::sndfile` and `fairseq2n::detail::kaldi_fbank_computer` follow RAII convention. They acquire a raw resource (in this case the handle returned by their C API), store it in an internal variable, and release that resource in their destructors. Since destructors of local variables are always called even if the function raises an exception, RAII classes guarantee that they don't leak the resources they hold.

Reply:
I created 2 new classes, one that handles png_ptr and info_ptr and another class that handles cinfo for jpeg decoding. Their destructors release those resources